src/mbgl/text/glyph_set.cpp

```diff
@@ -81,88 +80,168 @@ void align(Shaping& shaping,
-    //  tolerance for going a little
-    //  over maxWidth
-    for (uint32_t i = 0; i < logicalInput.size() - 1; i++) {
-        auto it = sdfs.find(logicalInput[i]);
-        if (it != sdfs.end())
+    for (char16_t chr : logicalInput) {
+        auto it = sdfs.find(chr);
+        if (it != sdfs.end()) {
             totalWidth += it->second.metrics.advance + spacing;
+        }
     }
 
-    int32_t lineCount = std::fmax(1, std::ceil(totalWidth / maxWidth));
-    return totalWidth / lineCount;
+    int32_t targetLineCount = std::fmax(1, std::ceil(totalWidth / maxWidth));
+    return totalWidth / targetLineCount;
+}
+    
+float calculateBadness(const float lineWidth, const float targetWidth, const float penalty, const bool isLastBreak) {
+    const float raggedness = std::pow(lineWidth - targetWidth, 2);
+    if (isLastBreak && lineWidth < targetWidth) {
+        // Be more tolerant of short final lines
+        return std::fmax(0, raggedness - 150);
+    }
+    if (penalty < 0) {
+        return raggedness - std::pow(penalty, 2);
+    }
+    return raggedness + std::pow(penalty, 2);
 }
+    
+float calculatePenalty(char16_t codePoint, char16_t previousCodePoint) {
+    float penalty = 0;
+    // Force break on newline
+    if (codePoint == 0x0a) {
+        penalty -= 10000;
+    }
+    // Penalize open parenthesis at end of line
+    if (previousCodePoint && (previousCodePoint == 0x28 || previousCodePoint == 0xff08)) {
+        penalty += 50;
+    }
+    // Penalize close parenthesis at beginning of line
+    if (codePoint == 0x29 || codePoint == 0xff09) {
+        penalty += 50;
+    }
+    return penalty;
+}
+    
+struct PotentialBreak {
+    PotentialBreak(const std::size_t p_index, const float p_x, const std::shared_ptr<PotentialBreak> p_priorBreak, const float p_badness)
+        : index(p_index), x(p_x), priorBreak(p_priorBreak), badness(p_badness)
+    {}
+    
+    const std::size_t index;
+    const float x;
+    const std::shared_ptr<PotentialBreak> priorBreak;
+    const float badness;
+};
+    
+struct LineBreakEdge {
+    LineBreakEdge(const std::shared_ptr<PotentialBreak> p_priorBreak, float p_badness)
+        : priorBreak(p_priorBreak), badness(p_badness)
+    {}
+    
+    std::shared_ptr<PotentialBreak> priorBreak;
+    float badness;
+};
+    
+std::shared_ptr<PotentialBreak> evaluateBreak(const std::size_t breakIndex, const float breakX, const float targetWidth, const std::vector<std::shared_ptr<PotentialBreak>>& potentialBreaks, const float penalty, const bool isLastBreak) {
+    // We could skip evaluating breaks where the line length (breakX - priorBreak.x) > maxWidth
+    //  ...but in fact we allow lines longer than maxWidth (if there's no break points)
+    //  ...and when targetWidth and maxWidth are close, strictly enforcing maxWidth can give
+    //     more lopsided results.
+    
+    std::vector<LineBreakEdge> edges;
+    // We're iterating from last potential break towards the first
```

Comment:
Is this iteration order necessary or could it use `for (const auto& potentialBreak : potentialBreaks)`?